# Integrative Genomics Workshop

### Melyssa Minto ([melyssa.minto\@duke.edu](mailto:melyssa.minto@duke.edu){.email})

Classes will be Sep 1st-2nd. Location TBA to prospective attendees and will most likely be hybrid. Email [melyssa.minto\@duke.edu](mailto:melyssa.minto@duke.edu){.email} to sign up, receive further updates, and ask questions about the workshop.

![](https://ars.els-cdn.com/content/image/1-s2.0-S0166223618300559-gr1.jpg)

## Learning Objectives

-   Basic biology of gene regulation and epigenetics
-   Introduction to differential RNA-seq analysis and differential binding analysis
-   Survey methods to map chromatin data to gene expression data
-   Implement peak to gene mapping using the programming languages R and Bash

## Overview

Each cell in our bodies comprises of the same DNA and yet this DNA gives rise to cells that vary widely in form and function. The human genome stretched out is about 2 meters long and many might find it surprising that only less than 3% of our genome encodes for genes. Initially, the remaining majority of our genome was thought of as "junk" DNA but we now know that these regions serves as hot beds for epigenomic regulation of gene expression. The study of epigenomic gene regulation includes investigating available chromatin, transcription factor binding, DNA methylation, and histone tail modifications. All of which contribute to the state of the chromatin and gene regulation. With the rise of cost-effective sequencing technologies to assay various forms of the epigenome, we can start to ask more questions regarding the molecular mechanism of gene activation and repression in many different contexts. However, a huge challenge in this field is linking the epigenomic state of intragenic regions to genes to infer the specific mechanisms that underlie the regulation of a particular gene or gene set. **In this workshop we will learn various methods to map regions with epigenomic to genes in a biologically informed way.** Each session will be 1.5 hrs. We will write R and bash/shell scripts execute all analyses. Some familiarity with programming is preferred but not necessary. You will need to bring your own laptop with RStudio installed or alternatively, use [Duke's RStudio VMs](https://shib.oit.duke.edu/idp/authn/external?conversation=e2s1). If you are new to programming and genomic analysis, this is a great resource to get some background:

## Detailed Schedule

| Day      | Topics                                                                                                   | 
|----------|----------------------------------------------------------------------------------------------------------|
| Sep 1st | [Intro to Epigenomics & Gene Regulation, Differential Analysis](exercises/IntroToEpigenomics.md) | 
| Sep 2nd | [Methods to map peaks to genes](exercises/Methods_PeakMapping.md)                                     |
