---
title: "Methods to Map Peak to Genes(Day 2)"
output: html_notebook
---

# Methods to Map Peak to Genes(Day 2)

### Learning outcomes

-   You will learn several types of peak to gene mapping methods

# Setting Up the framework

Once you know what peaks change the next step is to tie this data to gene expression or gene regulation. For example if you have a transcription factor peak increase in the promoter of gene_a and that gene_a increases in expression then you can associate that transcription factor with the up regulation of gene_a. Conversely, if that same transcription factor increasing in binding at gene_b and gene_b decreases in expression then that TF can be associated with down-regulation of gene_b.

✎ Assume you have count data from a ChIP-seq of TF peak case and control as well as gene expression values between case and control. What are the steps in identifying which genes the TF regulates between case and control? ✎

> 1.  Differential analysis of the Peak regions to determine where the TF gains binding and loses binding in comparison to the WT samples.
>
> 2.  Differential analysis of the gene expression to determine which genes are unregulated and down-regulated in comparison to the WT samples.
>
> 3.  Map the peaks to genes. Peaks that change that are mapped to genes that change can be associated with regulation of those genes

## Interpretation framework

✎Fill out this table to interpret peaks that change that are mapped to genes that change✎

|                     |             |           |
|---------------------|-------------|-----------|
|                     | Peak Gained | Peak Lost |
| Gene Up-regulated   |             |           |
| Gene Down-regulated |             |           |

\\

# Peak mapping methods

## Nearest gene

The naive method in mapping peaks to genes is simply mapping it to its nearest gene. You will know the genomic location of your peaks and you can use the TSS of each gene. Using those coordinates you can map peaks to the nearest gene. Luckily for us, there is a well documented R package that will do this for you.

The `ChIPseeker` package has many functions to work with chromatin data. In this course we will go through how to use the `annotatePeaks()` function in which you pass in an object that holds the peak positions (Chr, Start, Stop), and you specify the genome (mouse, human, fly) and it will output a nearest gene for each peak as well as some information about the peak in relation to the gene.

Before we run the code, let's check out the default options of `annotatePeaks()`

```{r load libraries}
library(ChIPseeker)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(tidyverse)
```

```{r}
?annotatePeak()
```

```{r}
# get path to peak bed file 
sample_file = "../workshop_data/output/down_peaks.bed"

```

```{r}
# run annotatePeak()
annot = annotatePeak(sample_file, 
                     TxDb=TxDb.Mmusculus.UCSC.mm10.knownGene, 
                     tssRegion=c(-1000, 1000), 
                     verbose=TRUE)


```

```{r}
# plot
plotAnnoBar(annot)
```

```{r}
vennpie(annot)

```

You can also do this with multiple bed files using the `lapply()` function,

```{r}
# get path to peak bed files 
sample_files = list.files("", pattern = "", full.names = TRUE)
sample_files <- as.list(sample_files)

```

```{r}
names(sample_files) <- c()

```

```{r}
# run annotatePeak()
zic_anno = lapply(sample_files, annotatePeak, TxDb=TxDb.Mmusculus.UCSC.mm10.knownGene, tssRegion=c(-1000, 1000), verbose=TRUE)

```

```{r}
# plot

```

✎ Brainstorm some questions you can explore with the data with peaks mapped to nearest gene✎

> **1.** What are the distribution of peaks mapped to genes that are expressed? or deferentially expressed?

### Explore the data

```{r}
str( zic_anno$Early@anno)
```

```{r}
# extract the annotation data from object
early_annotations = zic_anno

```

```{r}
# BiocManager::install("EnsDb.Mmusculus.v79")

library(EnsDb.Mmusculus.v79)

```

```{r ensmbl to symbols}
edb <- EnsDb.Mmusculus.v79
tx <- transcripts(edb, columns=c("tx_id", "gene_id", "gene_name"))
mapping <- data.frame(tx_id=tx$tx_id, SYMBOL=tx$gene_name)

head(mapping)
```

```{r}
# adding gene symbols
early_annotations = early_annotations %>% 
  mutate(tx_id = gsub("\\..*","",transcriptId)) %>% 
  left_join(mapping, by = "tx_id" ) 

head(early_annotations)
```

```{r diffexp}
# read in differential expression data in workshop folder
diff_exp = read_csv("")
```

```{r}
# add diff exp data to annotations
early_annotations = 

head(early_annotations)
```

### lets ask some questions

Are early peaks near early genes?

```{r}
# create a boxplot of the distiance to TSS for genes that are up-down regulated, down-regulated, and not significant

```

What kids of genes are early peaks near?

```{r}
# extract genes that zic bnds to promoters
early_near_genes = early_annotations %>% 
  dplyr::select(seqnames, start, end, SYMBOL, distanceToTSS, sig ) %>% 
  dplyr::filter(distanceToTSS < 3000) %>% 
  pull(SYMBOL)

head(early_near_genes, 20)
```

```{r}
library(clusterProfiler)
library(org.Mm.eg.db)
```

```{r, eval=FALSE}
# GO enrichment (DO NOT RUN)
go_bp = compareCluster(early_near_genes,
                       OrgDb = org.Mm.eg.db,
                       fun           = "enrichGO",
                       keyType = "SYMBOL",
                       pvalueCutoff  = 0.05,
                       pAdjustMethod = "BH",
                       ont = "BP")
```

```{r, eval=FALSE}
dotplot(go_bp)
```

## Using 3D Chromatin Information

So with the nearest gene methods, you may get peaks that map kilobases away from the gene. How can we be sure that that mapping is biologically relevant? Now, we can use 3D chromatin data to map genes to their enhancers and use this link as a means of mapping distal peaks to genes.

```{r read in loop data}

cerebellum_loops = read_tsv("../workshop_data/loops/combined_MAPS_peaks.txt", col_names = F)
head(cerebellum_loops)

```

Lets work through how we would do this, we want the data to look something like this:

+------------+----------------+-----------------+------------+--------------------+
| loop       | anchor1        | anchor2         | gene       | intersecting peaks |
+============+================+=================+============+====================+
| loop1      | chr1:1000-2000 | chr1:4000-5000  | geneA      | chr1:1150-1250     |
|            |                |                 |            |                    |
|            |                |                 |            | chr1:1500-1600     |
|            |                |                 |            |                    |
|            |                |                 |            | chr1:1750-1950     |
|            |                |                 |            |                    |
|            |                |                 |            | chr1:4200-4300     |
|            |                |                 |            |                    |
|            |                |                 |            | chr1:4500-4700     |
+------------+----------------+-----------------+------------+--------------------+
| loop2      | chr1:6000-7000 | chr1:9000-10000 | geneB      | chr1:6600-7000     |
|            |                |                 |            |                    |
|            |                |                 |            | chr1:9100-9200     |
|            |                |                 |            |                    |
|            |                |                 |            | chr1:9300-9400     |
|            |                |                 |            |                    |
|            |                |                 |            | chr1:9450-9900     |
+------------+----------------+-----------------+------------+--------------------+
|            |                |                 |            |                    |
+------------+----------------+-----------------+------------+--------------------+

What to we need to do

1.  **map loops to genes**

    1.  treat the loop anchors like peaks

    2.  map them to nearest gene using `annotatePeaks()`

    3.  filter loops for at list one of the anchors mapping to the promoter (`distanceToTSS =0`)

2.  **get peak intersection with loop anchors.** Use `bedtools` to get the intersection of peaks in anchors

    ![](https://bedtools.readthedocs.io/en/latest/_images/intersect-glyph.png)

3.  **map the peaks to gene via the loops.** For this we will just need to combine and reformat the data

    +--------------+-----------------+-----------------+-----------------+
    | loop         | anchors         | peaks           | gene            |
    +==============+=================+=================+=================+
    | loop1        | chr1:1000-2000  | chr1:1150-1250  |                 |
    +--------------+-----------------+-----------------+-----------------+
    | loop1        | chr1:1000-2000  | chr1:1500-1600  |                 |
    +--------------+-----------------+-----------------+-----------------+
    | loop1        | chr1:1000-2000  | chr1:1750-1950  |                 |
    +--------------+-----------------+-----------------+-----------------+
    | loop1        | chr1:4000-5000  | chr1:4200-4300  | geneA           |
    +--------------+-----------------+-----------------+-----------------+
    | loop1        | chr1:4000-5000  | chr1:4500-4700  | geneA           |
    +--------------+-----------------+-----------------+-----------------+

    LETS CODE IT:

```{r reformat loop data}

 



  
```



```{r write loop data}
loop_data 
```

### anchor peak instersection

[Bedtools intersect](https://bedtools.readthedocs.io/en/latest/content/tools/intersect.html)

```{bash}
# bedtools intersect loop anchors with zic peaks

```



### anchor gene mapping 
```{r loop to gene mapping}
# run annotate peak on loops
anchor_annot = annotatePeak()
```

```{r plot annotations}
```


```{r combine loop data with gene data}

loop_gene = anchor_annot@anno 
```

```{r combine loop data with peak data }
# read in peak-anchor intersections
early_peak_loop = read_tsv()
head(early_peak_loop)
```

```{r}
# reformat data to merge with loop data
early_peak_loop = 

head(early_peak_loop)
```

```{r}
# join peak-anchor and loop-ids
early_peak_loop = 

head(early_peak_loop)
```

```{r}
# combine all data
all_data = 
head(all_data)
```

Give it a try with the late peaks:
```{bash}

```

```{r}

```


```{r}

```


```{r}

```

### lets ask some questions

First lets add in the gene expression
```{r}
all_data = all_data %>% 
  left_join(diff_exp)

head(all_data)
```

How many genes are differentially expressed and mapped to peaks?
```{r}
table(all_data$sig)
```

