---
title: Intro to Epigenomics
---

# Intro to Epigenomics (Day1)

### Learning outcomes

-   You will learn several types of epigenomic regulation and data

-   You will understand how epigenomic sequencing data is processed for analysis

-   Practice generating reading in count data into R and performing differential peak analysis

# The central Dogma of Biology

![ℹ️ Adapted from Namec et al (2021) Nat Rev](https://media.springernature.com/full/springer-static/image/art%253A10.1038%252Fs41578-020-00238-z/MediaObjects/41578_2020_238_Fig1_HTML.png?as=webp)

\
The **central dogma of molecular biology** has long reigned as the transformation of DNA to RNA (transcription) and from RNA to Protein (translation). We now know that proteins can interact with DNA to modulate the production of RNAs via trans-regulation. Similarly, modifications to the DNA (i.e. DNA methylation, histone tail modifications) work via cis-regulation to modulate the production of RNAs. Further we will discuss different levels of epigenetic control of the transcription.\

# **What is Epigenomics?**

**Epigenomics** is the study of molecular elements that regulate gene expression. The word Epigenome roughly translates to "on top of the genome" with "Epi" meaning top. Put simply, its the study of what is on our DNA. Historically, this has been studied by examining various modes of genomic data such as chromatin accessibility, Transcription Factor (TF) binding, modifications to the DNA (i.e. DNA methylation, histone tail modifications), and the overlap 3D conformation of the genome. Each of these layers have some functional role underlying the ability for RNA PolII to bind to promoters and initiate transcription.

## **Chromatin Accessibility**

![Adapted from: A continuum of accessibility states broadly reflects the distribution of chromatin dynamics across the genome, Klemm et al, 2019](https://media.springernature.com/full/springer-static/image/art%253A10.1038%252Fs41576-018-0089-8/MediaObjects/41576_2018_89_Fig1_HTML.png?as=webp)

The DNA is wrapped around **nucleosomes** which then can wrap around itself like a tangled telephone cord which leads to a compact chromatin state. Chromatin compaction is thought to be one of the largest driving forces in transcript. Transcriptional machinery and other epigenetic factors simply cannot bind the the DNA if the chromatin is "closed" or in a compact **hetero-chromatin** state.

![Adapted from: Principal methods for measuring chromatin accessibility., Klemm et al, 2019](https://media.springernature.com/full/springer-static/image/art%253A10.1038%252Fs41576-018-0089-8/MediaObjects/41576_2018_89_Fig2_HTML.png?as=webp)

Recent and common methods to assay chromatin accessibility are ATAC-seq and DNase-seq.  Both methods are able to pull down regions of the genome that are more "open" and the sequencing readouts shows pileups or peaks of areas where the chromatin is accessible.

![Adapted from: Population-scale measurements of chromatin accessibility reflect the average accessibility of a heterogeneous collection of single molecules., Klemm et al. 2019](https://media.springernature.com/full/springer-static/image/art%253A10.1038%252Fs41576-018-0089-8/MediaObjects/41576_2018_89_Fig3_HTML.png?as=webp)

|        |                                                                                                                                                                 |
|--------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `NOTE` | Methods vary in the way they capture chromatin and careful consideration should be made in determining the method of choice and tools to best analyze the data. |

## **3D Functional Domains: TADs and LADs**

Even though the super coiling of a tangled telephone cord might seem random, the compaction and 3D organization of the DNA is not. ![Adapted from: Hierarchical organization of chromatin structure., Bonev & Cavalli, 2016](https://media.springernature.com/full/springer-static/image/art%253A10.1038%252Fnrg.2016.112/MediaObjects/41576_2016_Article_BFnrg2016112_Fig2_HTML.jpg?as=webp)

## Transcription Factors & Chromatin Remodelers

Proteins can interact with the DNA and have epigenomic control. **Transcription factors** (TFs) are a class of DNA binding proteins that bind in a sequence specific way. TFs are typically expressed in a precise spatiotemporal pattern when they are triggered by signalling cascades from biological cues like developmental triggers or a particular stimulus. TFs are known to bind to regulatory regions such as enhancers and promoters to activate and silence programs of gene expression.

**Chromatin remodelers** are proteins that interact with the DNA in an ATP-dependent manner to alter nucleosome positioning and thus the 3D architecture of the genome. These remodelers are known to form complexes that mediate histone turnover by inducing nycleosome sliding. This shifting of nucleosomes thus reveals sites for TF binding.

## DNA Methylation

Methyl groups can be added to bases of DNA by **Dnmt** proteins while **Tet** proteins are known to remove DNA methylation. DNA methylation is associated with gene silencing. DNA methylation occurs widely at **CpG sites**, that is where Cytosine and Guanine follow each other in a genomic sequence. However, **CpG Islands** are long stretches of DNA that is CG rich with less nucleosome density that is not typically heavily methylated. More recent research has shown that non CpG methylation is more common in neuronal tissue (specifically at CpA sites) and plays an important role in gene regulation.

![](https://www.mdpi.com/genes/genes-08-00148/article_deploy/html/images/genes-08-00148-g002-550.jpg)

## Histone Tail Modification

As DNA is wrapped around nucleosomes comprised of histone proteins, those histones have flexible tails that can be postranslationally modified which results is changes in chromatin conformation and accessibility. For instance, acetylation of Histone 3 Lysine at position 27 (H3K17ac) is associated with euchromatin, active enhancers and transcription and tri-methylation of Histone 3 at Lysine 27 is associated with hetero-chromatin and gene silencing. One can assay for different histone tail modifications via ChIP-sequencing. Similar to TF ChIP, one can align these reads to the genome to get pileups of reads and infer where specific modification occur in the genome. Some histone modifications tend to have a large spread while others are known to have discrete pins of binding resulting in sharp peaks. This should be considered when determining the algorithms and parameters used to determine loci from sequencing assays

✎ What is the difference between DNA Methylation and Histone Methylation? ✎

> DNA methylation occurs directly on the nucleotide and is typically associated with chromatin repression. Histone methylation occurs on the tails of histone. Specific residues of the histone tail can me methylated up to three times (mono-, di-, tri- methylation). Histone methylation is associated with varying chromatin and gene regulation depending on the residue that is methylated and the how many methyl groups are added (1-3).

# Raw Reads to count matrix

There are an increasing number of technologies to assay the different epigenomic marks. All of these technologies produce segments of sequences that was captured. Generally, the goal is to use algorithms to estimate where an epigenetic marks on the genome by taking the raw sequences, align them to a reference genome, and estimate a quantitative "count" of where those sequences tend to pileup.

![Adapted from ENCODE, Image credits: Darryl Leja (NHGRI), Ian Dunham (EBI), Michael Pazin (NHGRI)](https://www.encodeproject.org/images/c45f4d8c-0340-4fcb-abe3-e4ff0bb919be/@@download/attachment/EncodeDatatypes2013-7.png)

## Raw Sequencing data

![️ Adapted from Nakato and Sakata (2021) Methods](https://ars.els-cdn.com/content/image/1-s2.0-S1046202320300591-gr1.jpg)

Raw sequencing data is typically stored in a .fastq file and holds the raw sequencing reads as well as important information about each read including the quality of each base. Using that data, one can use an aligner (e.g. Bowtie2, STAR) to map the reads to a reference genome. Then one can use a peak caller (e.g. MAC2) to find where many reads pile up in the genome. Once you know where the peaks of reads are in the genome there are various down stream analyses:

1.  Differential peak analysis

2.  Annotation of the peaks by genomic loci or chromatin state

3.  Mapping peaks to genes to estimate which genes those marks may regulate.

In this course we will demonstrate a selection of methods on how to do differential peak analysis. The major steps include 1) calling peaks 2) generating the peak count matrix 3) analyzing count data 4) clustering peaks by count 5) Differential analysis of the peaks between two conditions

## Calling peaks

To call peaks, we will use [MACS2](https://pypi.org/project/MACS2/) a widely used tool to call peaks for ChIP-seq, ATAC-seq, and DNase-seq data. There are parameters such as `--broad` that can be used to tune the peak calling for epigenetics marks that may either sharp narrow peaks or broad-spread peaks.

Here is an example usage for calling ATAC peaks as suggested by `MACS2` documentation

``` bash
macs2 callpeaks
```

The peaks will likely be in some sort of .bed format with at minimum three columns representing the Chr Start and End positions of each peak.

``` bash
chr1 
```

Once peaks are called from each sample, the next step is to compare the peaks between samples or conditions. To do that, a consensus peak set must me made. This can be done in various ways from simply merging all the peaks into one file or using tools to statistically infer confident peaks from each set. `bedtools merge` is a common tool to merge peak sets.

\
![](https://bedtools.readthedocs.io/en/latest/_images/merge-glyph.png)

\
To merge peaks sets you need to first combine them into one file, sort it and then merge. One important step in this is to filter out regions of the genome that are highly mappable or that have a lot of repeats. This region is called the 'blacklist'. Common genomes have blacklists that can be found here:

<https://github.com/Boyle-Lab/Blacklist/tree/master/lists>

``` bash
cat peak_set1.bed peak_set2.bed peak_set3.bed > all_peaks.cat
bedtools sort all_peaks.cat > all_peaks.sort
bedtools merge all_peaks.sort > all_peaks.merge
bedtools filter all_peaks.merge blacklist.bed >  all_peaks.bed
```

|        |                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|--------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `Note` | The file extensions (eg .cat, .sort) are not specific to any program. Here I use them to help denote what step in merging process each file is. All the files that are made during this process are technically still .bed files which are tab delimited files text files with specific guidelines about what goes in each column. See more about the different .bed files here: <http://useast.ensembl.org/info/website/upload/bed.html> |

✎ What does the `cat` function do? ✎

> `cat` is short for con**cat**enate which simply means to combine together in a series. For example,
>
> ``` bash
> cat "string1" "string2"
> ```
>
> Will output in
>
> ``` bash
> string1string2
> ```

------------------------------------------------------------------------

### Lets learn a little about this data:

-   Study: [Regulation of chromatin accessibility and Zic binding at enhancers in the developing cerebellum](https://pubmed.ncbi.nlm.nih.gov/25849986/)

-   System: Developing mouse cerebellum at timpoints postnatal day 7 (early) and postnatal day 60 (late)

-   Major findings

    -   Chromatin accessibility changes throughout the developing cerebellum

    -   The family of Zic TF motifs enriched in sites of open chromatin

    -   Zics - Zinc Finger TFs of the cerebellum

    -   It was surprising to see them at the late stages because they were only thought of as an early neurogenesis factor not one that was involved in maturing neurons

    -   The gene expression of Zic1/2 are constitutive BUT the binding of Zic1/2 protein are highly dynamic between Early and Late cerebellar development

```{bash call peaks}
# the directory where the called peaks are
DATADIR=../workshop_data/peaks/

# make a sub directory for output
mkdir ../workshop_data/peaks/merged_peaks/
SAVEDIR=../workshop_data/peaks/merged_peaks/


# concatenate all peaks into one file
cat $DATADIR*.narrowPeak > $SAVEDIR/all_peaks.cat
# cort the peaks by cooordinates
bedtools sort -i $SAVEDIR/all_peaks.cat > $SAVEDIR/all_peaks.sort
# merge overlapping peaks
bedtools merge -i $SAVEDIR/all_peaks.sort > $SAVEDIR/all_peaks.merge
# remove regions that are blacklisted
bedtools subtract -a $SAVEDIR/all_peaks.merge -b ../workshop_data/mm10-blacklist.bed > $SAVEDIR/all_peaks.bed
```

```{bash call peaks}
# take a look at file outputs
ls -lh ../workshop_data/peaks/merged_peaks/
```

## Generate Peak Count Matrix

Once there is a consensus list of regions to compare, the next step is to calculate the number of reads (or counts) from these regions in each sample. This is usually organized in a table where each column is a sample and each region (or peak) is a row.

| peak  | Sample1 | Sample2 | Sample3 | Sample4 |
|-------|---------|---------|---------|---------|
| peak1 | 30      | 24      | 15      | 10      |
| peak2 | 45      | 60      | 22      | 18      |
| peak3 | 0       | 1       | 0       | 0       |
| peak4 | 127     | 88      | 90      | 98      |
| ...   | ...     | ...     | ...     | ...     |

This can easily be done using the R package `Rsubreads`

```{r load libraries}
# install packages
# install.packages("readr")
# install.packages("dplyr")
# BiocManager::install("Rsubread")

# load them
library(readr)
library(dplyr)
library(Rsubread)
```

```{r read in files}
# read in the consensus peak set
peak_set <- read_tsv("../workshop_data/peaks/merged_peaks/all_peaks.bed", col_names = F)
head(peak_set)

```

```{r}
# create a vector with the paths for all the bam files
bam_paths = list.files("../workshop_data/bams", pattern = ".bam", full.names = T)
bam_paths
```

✎ How can you learn more about a specific R function? ✎

> 1.  You can google! 🔍 someFunction() R
> 2.  Use the R help function within the console: \> ?someFunction()
> 3.  Search for the function in the Packages pane in Rstudio
>
> <img src="https://confluence.wustl.edu/download/attachments/189860067/rstudio-packages.png?version=1&amp;modificationDate=1639065252927&amp;api=v2" alt="R Studio Packages Pan" height="300"/>
>
> Good functions tend to have detailed descriptions of their arguments and some nice examples to model your code off of. It is generally good practice to read over the different arguments ( especially the defaults) that each function has to ensure that it is optimized to your data.

```{r}
?featureCounts
```

```{r Get Peak Counts}
# formatting the consensus peak set
peak_set = peak_set %>%
mutate(GeneID = paste0("peak", 1:n()),
       Chr = X1,
       Start = X2,
       End = X3,
       Strand = ".") %>%
select(GeneID, Chr, Start, End, Strand)

head(peak_set)
```

```{r Get Peak Counts}
# run function to get counts
peak_counts = featureCounts(files = bam_paths,
                            annot.ext  = peak_set,
                            nthreads = 2,
                            isPaired = FALSE)
```

```{r}
# check results 
str(peak_counts)
```

```{r}
head(peak_counts$counts)
```

## Analyze count data

Now that there is a count matrix we can ask different questions with the data.

1.  How do the samples cluster?

2.  Where are these peaks in the genome?

3.  Which Peaks are different from each other?

4.  Are there any patterns in the peaks that are change that we can see?

### Clustering

A good quality assurance check when working with count data, especially between different conditions is to check how well conditions cluster together. This is commonly done using a PCA. PCA is unsupervised modeling technique that decreases the number of dimensions in the data. The first principle component (PC1) is a line/plane in the data that explains of the highest variation in that data. The second Principle Component (PC2) is perpendicular to PC1 that explains the 2nd highest variation in the data. In sequencing analysis PCA is used to plot the top variation in the data to show that conditions cluster together. To learn more watch this [video](https://www.youtube.com/watch?v=FgakZw6K1QQ).

![](https://chrisalbon.com/images/machine_learning_flashcards/Principal_Component_Analysis_print.png)

```{r load library for count data}
library(DESeq2)
library(tidyr)
library(ggplot2)

```

```{r set up deseeq object}

counts <- as.data.frame(peak_counts$counts) 
colnames(counts) = c("early_1", "early_2", "late_1", "late_2")

design_data = data.frame( sample = c("early_1", "early_2", "late_1", "late_2"),
                          group = c("early", "early", "late", "late")
)

head(counts)
head(design_data)
```

```{r}
#create deseq object DESeq2
dds<-DESeqDataSetFromMatrix(countData = counts, 
                            colData = design_data,
                            design = ~ group)
```

```{r}
?plotPCA
```

```{r}

# extract normalized reads by calculating the variance stabilized counts
norm_reads <- vst(dds)
head(norm_reads)
```

```{r}
plotPCA(object = norm_reads,intgroup=c("group")) +
  theme_classic()
```

### Differential Analysis

Ultimately, most researchers want to know what the gene expression or binding differences are between two or more groups. There are many tools that can carry out differential analysis which vary in statistical approaches. The typical pipeline includes 1) getting the counts 2) filtering and normalization 3) statistical testing 4) select significant features. A large part of what varies between tools is 1) the underlying assumptions about how the count data is distributed and 2) how they model feature-wise and sample-wise variance/dispersion. This workshop will focus on the use of `DESeq2` but there are other tools like `diffbind` that are worth looking into.

With `DESeq2` it is recommended that you input raw unnormalized counts. The tool will compute a log-scaled normalization and filtering based on the distribution of the average count value of each feature. In our case our features are peaks and our columns are samples.

```{r perform analysis}
dds<-DESeq(dds)

```

```{r extract results}
# format results
res = results(dds, contrast = c("group", "late", "early"), tidy = TRUE)
head(res)

```

```{r}
res = res %>% 
  dplyr::rename(GeneID = row) %>% 
  left_join(peak_set) %>% 
  dplyr::select(-Strand) %>% 
  dplyr::mutate(sig = case_when(log2FoldChange >= 1 & padj < 0.05 ~ "UP",
                                     log2FoldChange <= 1 & padj < 0.05 ~ "DOWN",
                                     is.na(padj) ~ "filtered",
                                     TRUE ~ "N.S.")) %>% 
  dplyr::rename(PeakID = GeneID)

head(res)
```

```{r visualization}
# MA plot
plotMA(dds)
```

```{r}
# lets build it out from scratch
res %>% 
  mutate(colrs = ifelse(sig %in% c("UP", "DOWN"), "yes", "no")) %>% 
  ggplot(aes(x = baseMean, y = log2FoldChange, color = colrs)) +
  geom_point() +
  scale_color_manual(limits = c("yes", "no"), values = c("blue", "grey")) +
  scale_x_log10()+
  theme_bw() +
  labs( x = "mean of normalized counts", y = "log fold change")+
  theme(legend.position = "none") 
```

### Output Differentially Accessible peaks

```{r }
up_peaks = res %>% 
  filter(sig == "UP") %>% 
  select(Chr, Start, End)

head(up_peaks)
```

```{r}
down_peaks = res %>% 
  filter(sig == "DOWN") %>% 
  select(Chr, Start, End)

head(down_peaks)
```

```{r write}
down_peaks %>% write_tsv("../workshop_data/output/down_peaks.bed", col_names = F)
up_peaks %>% write_tsv("../workshop_data/output/up_peaks.bed", col_names = F)


```

# Glossary

|                      |                                                                                                                            |
|----------------------|----------------------------------------------------------------------------------------------------------------------------|
| **Nucleosome**       | unit of chromatin that is comprised of 147 bps of DNA wrapped around an octomer of histone proteins (H2A, H2B, H3, and H4) |
| **Hetero-chromatin** |                                                                                                                            |
| **Eu-chromatin**     |                                                                                                                            |

------------------------------------------------------------------------

# References

-   [Nemec, S., Kilian, K.A. Materials control of the epigenetics underlying cell plasticity. Nat Rev Mater 6, 69--83 (2021).](https://doi.org/10.1038/s41578-020-00238-z)
-   [Klemm, S.L., Shipony, Z. & Greenleaf, W.J. Chromatin accessibility and the regulatory epigenome. Nat Rev Genet 20, 207--220 (2019).](https://doi.org/10.1038/s41576-018-0089-8)
-   [Bonev, B., Cavalli, G. Organization and function of the 3D genome. Nat Rev Genet 17, 661--678 (2016).](https://doi.org/10.1038/nrg.2016.112)
-   [Moore, L., Le, T. & Fan, G. DNA Methylation and Its Basic Function. Neuropsychopharmacol 38, 23--38 (2013).](https://doi.org/10.1038/npp.2012.112)
-   [Jang HS, Shin WJ, Lee JE, Do JT. CpG and Non-CpG Methylation in Epigenetic Gene Regulation and Brain Function. Genes. 2017; 8(6):148.](https://doi.org/10.3390/genes8060148)
-   [West AE, Greenberg M. Neuronal; Activity-Related Gene Transcription in Synapse Development and Cognitive Function](https://cshperspectives.cshlp.org/content/3/6/a005744)
-   [Nakato R, Sakata T. Methods for ChIP-seq analysis: A practical workflow and advanced applications. Methods. 2021 Mar;187:44-53. Epub 2020 Mar 30.](https://www.sciencedirect.com/science/article/pii/S1046202320300591)
